server/src/session.rs

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) ZeroC, Inc.
+
+use crate::configuration_set::ConfigurationSet;
+use tokio::sync::{Mutex, RwLock};
+use tower_lsp::lsp_types::{ConfigurationItem, DidChangeConfigurationParams, Url};
+
+pub struct Session {
+    client: tower_lsp::Client,
+    /// This vector contains all of the configuration sets for the language server. Each element is a tuple containing
+    /// `SliceConfig` and `CompilationState`. The `SliceConfig` is used to determine which configuration set to use when
+    /// publishing diagnostics. The `CompilationState` is used to retrieve the diagnostics for a given file.
+    pub configuration_sets: Mutex<Vec<ConfigurationSet>>,
+    /// This is the root URI of the workspace. It is used to resolve relative paths in the configuration.
+    pub root_uri: RwLock<Option<Url>>,
+    /// This is the path to the built-in Slice files that are included with the extension.
+    pub built_in_slice_path: RwLock<String>,
+}
+
+impl Session {
+    pub fn new(client: tower_lsp::Client) -> Self {
+        Self {
+            client,
+            configuration_sets: Mutex::new(Vec::new()),
+            root_uri: RwLock::new(None),
+            built_in_slice_path: RwLock::new(String::new()),
+        }
+    }
+
+    // Update the properties of the session from `InitializeParams`
+    pub async fn update_from_initialize_params(
+        &self,
+        params: tower_lsp::lsp_types::InitializeParams,
+    ) {
+        // This is the path to the built-in Slice files that are included with the extension. It should always
+        // be present.
+        let built_in_slice_path = params
+            .initialization_options
+            .and_then(|opts| opts.get("builtInSlicePath").cloned())
+            .and_then(|v| v.as_str().map(str::to_owned))
+            .expect("builtInSlicePath not found in initialization options");
+        *self.built_in_slice_path.write().await = built_in_slice_path;
+
+        // Use the root_uri if it exists temporarily as we cannot access configuration until
+        // after initialization. Additionally, LSP may provide the windows path with escaping or a lowercase
+        // drive letter. To fix this, we convert the path to a URL and then back to a path.
+        let root_uri = params
+            .root_uri
+            .and_then(|uri| uri.to_file_path().ok())
+            .and_then(|path| Url::from_file_path(path).ok())
+            .expect("root_uri not found in initialization parameters");
+        *self.root_uri.write().await = Some(root_uri);
+    }
+
+    // Update the stored configuration sets by fetching them from the client.
+    pub async fn fetch_configurations(&self) {
+        let built_in_path = &self.built_in_slice_path.read().await;
+        let root_uri_guard = self.root_uri.read().await;
+        let root_uri = (*root_uri_guard).clone().expect("root_uri not set");
```

Comment:
It's unlikely to happen, since this is only called during initialized, but there's a clear deadlock here.
In this function we lock `built_in_slice_path`, then `root_uri`.

... continued